test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+		if err != nil {
+			return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", outOfSpecMachine.Name())
+		}
+
+		if err := externalMachine.Delete(ctx); err != nil {
+			return errors.Wrapf(err, "failed to delete machine %s", outOfSpecMachine.Name())
+		}
+
+		outOfSpecMachineNames[outOfSpecMachine.Name()] = nil
+	}
+
+	var stats []*infrav1exp.DockerMachinePoolInstanceStatus
+	for _, machineStatus := range np.dockerMachinePool.Status.Instances {
+		if _, ok := outOfSpecMachineNames[machineStatus.InstanceName]; !ok {
+			stats = append(stats, machineStatus)
+		}
+	}
+
+	for _, overprovisioned := range stats[*np.machinePool.Spec.Replicas:] {
+		externalMachine, err := docker.NewMachine(np.cluster.Name, overprovisioned.InstanceName, np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", overprovisioned.InstanceName)
+		}
+
+		if err := externalMachine.Delete(ctx); err != nil {
+			return errors.Wrapf(err, "failed to delete machine %s", overprovisioned.InstanceName)
+		}
+	}
+
+	np.dockerMachinePool.Status.Instances = stats[:*np.machinePool.Spec.Replicas]
+	return np.Refresh()
+}
+
+func (np *nodePool) Delete(ctx context.Context) error {
+	for _, machine := range np.machines {
+		externalMachine, err := docker.NewMachine(np.cluster.Name, machine.Name(), np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", machine.Name())
+		}
+
+		if err := externalMachine.Delete(ctx); err != nil {
+			return errors.Wrapf(err, "failed to delete machine %s", machine.Name())
+		}
+	}
+
+	return nil
+}
+
+func (np *nodePool) reconcileMachine(ctx context.Context, machine *docker.Machine) error {
+	machineStatus := getInstanceStatusByMachineName(np.dockerMachinePool, machine.Name())
+	if machineStatus == nil {
+		machineStatus = &infrav1exp.DockerMachinePoolInstanceStatus{
+			InstanceName: machine.Name(),
+			Version:      np.machinePool.Spec.Template.Spec.Version,
+		}
+		np.dockerMachinePool.Status.Instances = append(np.dockerMachinePool.Status.Instances, machineStatus)
+	}
+
+	externalMachine, err := docker.NewMachine(np.cluster.Name, machine.Name(), np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", machine.Name())
+	}
+
+	if err := externalMachine.PreloadLoadImages(ctx, np.dockerMachinePool.Spec.Template.PreLoadImages); err != nil {
```

Comment:
am I wrong or are we going to reload images at every reconciliation?
what about doing this only if the machine is not yet bootstrapped?

Reply:
as alternative, we can exit earlier from this func if the machine is ready/if ProviderID is set

Reply:
moved inside of bootstrap conditional